packages/components/src/utils/css-variables.tsx

```diff
@@ -0,0 +1,201 @@
+	 * have been replaced with their computed values.
+	 *
+	 * Should be memoized to avoid unnecessary reflows.
+	 */
+	onChange: ( args: {
+		replacedCssString: string;
+		computedVariables: ComputedCSSVariables;
+	} ) => void;
+};
+type ComputedCSSVariables = Record< string, string >;
+type VarFunction = {
+	/** Start index of match. */
+	start: number;
+	/** End index of match. */
+	end: number;
+	/** Matched string, e.g. `var( --foo, #000 )`. */
+	raw: string;
+	/** CSS variable name, e.g. `--foo`. */
+	value: string;
+	/** Second argument of the `var()`, which could be a literal or another `var()`. */
+	fallback?: string;
+};
+
+/**
+ * Find the index of the matching closing parenthesis for a given opening parenthesis in a string.
+ */
+function findMatchingParenthesis( str: string, startPos: number ) {
+	let stack = 0;
+
+	for ( let i = startPos; i < str.length; i++ ) {
+		if ( str[ i ] === '(' ) {
+			stack++;
+		} else if ( str[ i ] === ')' ) {
+			stack--;
+			if ( stack === 0 ) {
+				return i;
+			}
+		}
+	}
+
+	throw new Error( 'No matching closing parenthesis found.' );
+}
+
+/**
+ * Find all `var()` functions in a CSS string.
+ */
+export function findVarFunctionsInString( str: string ) {
+	const regex = /(?<=\bvar)\(/g;
+	const matches: VarFunction[] = [];
+
+	let openingParen;
+	while ( ( openingParen = regex.exec( str ) ) !== null ) {
+		const closingParen = findMatchingParenthesis( str, openingParen.index );
+		const [ start, end ] = [
+			openingParen.index - 'var'.length,
+			closingParen + 1,
+		];
+		const raw = str.slice( start, end );
+		const value = raw.match( /--[\w-]+/ )?.[ 0 ];
+
+		if ( ! value ) {
+			throw new Error( 'No CSS variable found in var() function.' );
+		}
+
+		matches.push( {
+			start,
+			end,
+			raw,
+			value,
+			fallback: raw.match( /,(.+)\)/ )?.[ 1 ].trim(),
+		} );
+
+		regex.lastIndex = closingParen + 1; // resume next regex search after this closing parenthesis
+	}
+
+	return matches;
+}
+
+/**
+ * Get the computed CSS variables for a given element.
+ */
+function getComputedCSSVariables(
+	propertyStrings: string[],
+	element: HTMLDivElement
+) {
+	return propertyStrings.reduce( ( acc, propertyString ) => {
+		acc[ propertyString ] = window
+			.getComputedStyle( element )
+			.getPropertyValue( propertyString );
+		return acc;
+	}, {} as ComputedCSSVariables );
+}
+
+/**
+ * Replace all `var()` functions in a CSS string with their computed values.
+ */
+export function replaceCSSVariablesInString(
+	str: string,
+	computedVariables: ComputedCSSVariables
+): string {
+	const varFunctions = findVarFunctionsInString( str );
+
+	let result = '';
+	let lastIndex = 0;
+
+	varFunctions.forEach( ( { start, end, value, fallback } ) => {
+		const replacement =
+			computedVariables[ value ] ||
+			replaceCSSVariablesInString( fallback ?? '', computedVariables );
+
+		if ( ! replacement ) {
+			throw new Error( `No value found for CSS variable ${ value }.` );
```

Comment:
If this function is part of the public API, throwing an error can be a bit invasive. I see we're throwing it here to catch it in the `CSSVariableReplacer` but it may not be ideal if someone tries to consume `replaceCSSVariablesInString` directly and it errors just like that.